### ISSUE TITLE
Improved interface for optimizer.

### ISSUE THREAD

**Issue by ppwwyyxx** [ppwwyyxx/tensorpack#137]
In general, anything which shares the `tf.train.Optimizer` interface and "modifies variables" can be an optimizer. As an optimizer can be tightly bound to the graph, it's better to make it part of the model, not trainer.
Several design goodness:

1. it's not unusual if you want to use more than one optimizers. Having an single "optimizer" key in TrainConfig is not a very general thing. It's better to put it in models somehow and be something like ModelDesc.get_cost: you don't have to set it if you don't use it.
2. It's common to use an optimizer which depends on tensors/variables (e.g. VariableClippingOptimizer in tf.contrib). However currently the optimizer is initialized before the model. Having a Model.get_optimizer() method can solve this.
3. get_gradient_processor is a confusing thing. It seems better to explain it as a new kind of optimizer, which applies a function on the gradients.

**Comment by PatWie** [comment]
My understanding is, that the optimizer contains all code for an update-step. And the trainer feeds data into the model, call the optimizer and manage gpus.

So, isn't this:
https://github.com/ppwwyyxx/tensorpack/blob/master/examples/GAN/WGAN-CelebA.py#L70-L87
part of the optimizer itself?

**Comment by ppwwyyxx** [comment]
Do you mean putting the ` for k in range(5):` logic into optimizer?

**Comment by PatWie** [comment]
That would be the way if I would have tried to implement the changes. So the trainer just handles sync-multi-gpu, or so and each of these trainers can be applied to GANOptimizer, WGANOptimizer. Otherwise, whta is the advantage of putting the optimizer into the model, when it is just RMSprop, Adam and co?

Would be nice to hear your definition of an "optimizer" vs "trainer".

**Comment by ppwwyyxx** [comment]
The clipping is added to the optimizer to avoid having a call like "optimizer.post_update". It's more efficient if everything is done in one op. It's also why GAN example only have one train_op.
But there will always be some logic cannot be easily handled by an op, like scheduling two minimization ops in the WGAN case. 

I don't have a clear boundary between the two. Writing WGAN without the clipping optimizer (just apply the clipping manually) is also fine to me. I added the clipping optimizer because it is common (so there should be a tool for it), not because it shouldn't be in trainer.
What I had in mind is that: optimizers are just simple extensions of `tf.train.Optimizer` and should be mostly compatible with it (so it should produce an op in the end). 
Trainer does everything necessary to define what `run_step` is. It can use a slightly better optimizer if that's helpful, but in general all the non-standard logic should appear here.

**Comment by ppwwyyxx** [comment]
It may also be a good idea to move the python-side logic to "optimizer" (a new kind of optimizer incompatible with tf.train.Optimizer), so that trainer only handles the graph. I mentioned this because currently in GAN `_setup()` both build the graph and define the optimization ops.

But it's still not clear what is the boundary. Is multi-GPU training a kind of optimizer (because all its logic is to play with gradient stuff and make the update)? 

I still found it very hard to separate the two. So maybe I'll still let trainer do everything.. what do you think?

**Comment by PatWie** [comment]
I was in the middle of nowhere for some days with bad internet connection and missed several commits of tensorpack.

I think the trainer is related to the infra-structure: multi-gpu, multi-device, sync, asyn, QueueInput, TensorInput and a good optimizer should not know which hardware is used or where the data is coming from. An optimizer should provide a `compute_gradients` or `train_op`.
In the ideal case any GANoptimizer, WGANoptimizer should work under all Trainers, which is currently not the case.

So clipping and run multiple steps can be part of the optimizer itself, like you [already did](https://github.com/ppwwyyxx/tensorpack/blob/master/examples/GAN/WGAN-CelebA.py#L40-L51).

I would argue that the ratio to two optimization ops in the GAN should be part of the model. I will let you know if the is a discussable alternative, as the current implementation is understandable.

**Comment by ppwwyyxx** [comment]
It would be very ideal to use WGAN under multiGPU automatically, but I don't how I could do that. 

The boundary is still not clear in what an optimizer should do.
SyncMultiGPUTrainer produces one train_op in the end, should it use optimizer to implement the logic? AsyncMultiGPUTrainer uses several different train_ops, should it be in trainer?

There are ways to put the ratio into the graph with some loop operators, but it'll probably look very complicated. And in general users shouldn't be forbidden writing python loops inside run_step.
